### ISSUE TITLE
Xml serialization time grows exponentially with size of map

### ISSUE THREAD

**Issue by nadddy** [clbanning/mxj#70]
map.Xml() time grows exponentially for large map.
For example if we read 1MB XML and serialize it back it take almost 20s and if we double the size (duplicate the xml nodes from same input) then time taken is almost 90s.

I tested with this xml (http://aiweb.cs.washington.edu/research/projects/xmltk/xmldata/data/mondial/mondial-3.0.xml) with following code.

```
package main

import (
	"fmt"
	"io/ioutil"
	"time"

	mxj "github.com/clbanning/mxj"
)

func main() {
	b, _ := ioutil.ReadFile("/tmp/large.xml")
	start := time.Now()
	m, _ := mxj.NewMapXml(b)
	fmt.Printf("XML Read time %v\n", time.Now().Sub(start))
	start = time.Now()
	b, _ = m.Xml()
	fmt.Printf("XML Write time %v\n", time.Now().Sub(start))
}
```

I debugged further and I think it is because of string concatenation (*s += ). I created a simple code to string concatenation and found that it is very slow as string grows. When I replaced it by bytes.Buffer.Read then there was significant improvement.

**Comment by clbanning** [comment]
Please give me machine spec.  Just you program on my Mac Book Pro 2.6 GHz 6-Core Intel Core i7 w/ 16 GB and had time of 8.6s for the referenced XML file.

**Comment by nadddy** [comment]
Hi @clbanning 
Thanks for quick reply and thank for this tool, it is very helpful.

Please try doubling the xml and you will observe that time now will be in order of 30s on same machine.

Following is my simple script to compare string concat vs byte buffer:
```
package main

import (
	"bytes"
	"fmt"
	"time"
)

func main() {
	n := 100000
	s := new(string)
	b := new(bytes.Buffer)
	repeat := "repeat me"

	// string concatanation
	start := time.Now()
	for i := 0; i < n; i++ {
		*s += repeat
	}
	fmt.Printf("size = %v, duration = %v\n", len(*s), time.Now().Sub(start))

	// byte buffer
	start = time.Now()
	for i := 0; i < n; i++ {
		b.Write([]byte(repeat))
	}
	data := make([]byte, b.Len())
	b.Read(data)
	fmt.Printf("size = %v, duration = %v\n", len(data), time.Now().Sub(start))
}
```

And following is the output on my machine. Huge difference on same machine.
go run perf.go 
size = 900000, duration = 15.14816976s
size = 900000, duration = 6.377721ms

If you double n then the difference gorws even more:
go run perf.go 
size = 1800000, duration = 52.429814689s
size = 1800000, duration = 9.977366ms

**Comment by moorereason** [comment]
I thought a real benchmark comparing the loading of both files would be helpful:

```
$ benchstat bench-issue70-1.out bench-issue70-2.out
name       old time/op    new time/op    delta
Issue70-2    90.5ms ± 8%   224.8ms ±13%  +148.40%  (p=0.008 n=5+5)

name       old alloc/op   new alloc/op   delta
Issue70-2    30.0MB ± 0%    60.4MB ± 0%  +101.17%  (p=0.008 n=5+5)

name       old allocs/op  new allocs/op  delta
Issue70-2      608k ± 0%     1215k ± 0%   +99.99%  (p=0.008 n=5+5)
```

## Reproducing

```go
package mxj

import (
        "io/ioutil"
        "testing"
)

func BenchmarkIssue70(b *testing.B) {
        data, err := ioutil.ReadFile("issue70-1.xml")
        // data, err := ioutil.ReadFile("issue70-2.xml")
        if err != nil {
                b.Fatal("err:", err)
        }

        for i := 0; i < b.N; i++ {
                if _, err = NewMapXml(data); err != nil {
                        b.Fatal("err:", err)
                }
        }
}
```
I saved the `mondial-3.0.xml` file as `issue70-1.xml`, duplicated the contents inside the root `mondial` tag, and saved as `issue70-2.xml`.

Run against `issue70-1.xml`:
```
$ go test -run=XXX -bench=BenchmarkIssue70 -benchmem -count=5 | tee bench-issue70-1.out
goos: linux
goarch: amd64
pkg: github.com/clbanning/mxj
BenchmarkIssue70-2            12          88842484 ns/op        30033327 B/op     607617 allocs/op
BenchmarkIssue70-2            12          87380392 ns/op        30033264 B/op     607616 allocs/op
BenchmarkIssue70-2            12          87866634 ns/op        30035123 B/op     607623 allocs/op
BenchmarkIssue70-2            13          90714086 ns/op        30022817 B/op     607620 allocs/op
BenchmarkIssue70-2            13          97766087 ns/op        30022261 B/op     607618 allocs/op
PASS
ok      github.com/clbanning/mxj        7.033s
```

Run against doubled `issue70-2.xml`:
```
$ go test -run=XXX -bench=BenchmarkIssue70 -benchmem -count=5 | tee bench-issue70-2.out
goos: linux
goarch: amd64
pkg: github.com/clbanning/mxj
BenchmarkIssue70-2             6         235923807 ns/op        60364744 B/op    1215157 allocs/op
BenchmarkIssue70-2             6         197082870 ns/op        60365553 B/op    1215160 allocs/op
BenchmarkIssue70-2             6         219835822 ns/op        60362756 B/op    1215150 allocs/op
BenchmarkIssue70-2             5         253486245 ns/op        60481000 B/op    1215148 allocs/op
BenchmarkIssue70-2             5         217848762 ns/op        60482756 B/op    1215154 allocs/op
PASS
ok      github.com/clbanning/mxj        10.936s
```

**Comment by nadddy** [comment]
Hi @moorereason 

Is this benchmark for reading the data (xml to map). My issue is about serialization of map using map.Xml(). Reading I observed was very linear.

**Comment by moorereason** [comment]
@nadddy,
Sorry, I missed that part.  Here's the new results:

```
$ benchstat bench-issue70-1-xml.out bench-issue70-2-xml.out
name       old time/op    new time/op    delta
Issue70-2     23.1s ±10%     92.8s ± 5%  +301.86%  (p=0.008 n=5+5)

name       old alloc/op   new alloc/op   delta
Issue70-2    61.2GB ± 0%   244.2GB ± 0%  +299.14%  (p=0.008 n=5+5)

name       old allocs/op  new allocs/op  delta
Issue70-2      172k ± 0%      345k ± 0%  +100.25%  (p=0.008 n=5+5)
```

The test:
```go
func BenchmarkIssue70(b *testing.B) {
        data, err := ioutil.ReadFile("issue70-2.xml")
        if err != nil {
                b.Fatal("err:", err)
        }

        m, err := NewMapXml(data)
        if err != nil {
                b.Fatal("err:", err)
        }

        b.ResetTimer()
        for i := 0; i < b.N; i++ {
                if _, err = m.Xml(); err != nil {
                        b.Fatal("err:", err)
                }
        }
}
```

**Comment by moorereason** [comment]
I changed `mapToXmlIndent` to accept a `*strings.Builder` instead of `*string`, and these are the results:

```
$ benchstat bench-issue70-1-builder.out bench-issue70-2-builder.out
name       old time/op    new time/op    delta
Issue70-2    23.3ms ± 5%    48.1ms ±11%  +106.35%  (p=0.008 n=5+5)

name       old alloc/op   new alloc/op   delta
Issue70-2    11.0MB ± 0%    22.0MB ± 0%   +99.12%  (p=0.008 n=5+5)

name       old allocs/op  new allocs/op  delta
Issue70-2     65.6k ± 0%    131.1k ± 0%   +99.93%  (p=0.016 n=5+4)
```

Compared to `master`:
```
$ benchstat bench-issue70-1-xml.out bench-issue70-1-builder.out
name       old time/op    new time/op    delta
Issue70-2     23.1s ±10%      0.0s ± 5%  -99.90%  (p=0.008 n=5+5)

name       old alloc/op   new alloc/op   delta
Issue70-2    61.2GB ± 0%     0.0GB ± 0%  -99.98%  (p=0.008 n=5+5)

name       old allocs/op  new allocs/op  delta
Issue70-2      172k ± 0%       66k ± 0%  -61.95%  (p=0.008 n=5+5)
```

Happy to submit a PR if wanted.

**Comment by nadddy** [comment]
@moorereason 

Yes please submit a PR.
I have also changed using bytes.Buffer and got similar result but strings.builder looks better change. 
```
~/golib/bin/benchstat bench-issue70-master.out bench-issue70-branch.out 
name                         old time/op    new time/op    delta
LargeXmlUnmarshalMarshal-32     22.0s ± 5%      0.0s ± 2%  -99.87%  (p=0.008 n=5+5)

name                         old alloc/op   new alloc/op   delta
LargeXmlUnmarshalMarshal-32    61.2GB ± 0%     0.0GB ± 0%  -99.99%  (p=0.008 n=5+5)

name                         old allocs/op  new allocs/op  delta
LargeXmlUnmarshalMarshal-32      176k ± 0%       65k ± 0%  -63.01%  (p=0.016 n=5+4)
```

**Comment by clbanning** [comment]
So I've hacked a new version of m.Xml() and am getting the following metrics from your original post.
```
XML size 1784825
XML Read time 86.257641ms
XML Write time 14.932974ms
len of XML: 1156096
```

Got to wrap up some other stuff.  But will get it refactored into the package in the next day or so.  (If you need ASAP I can push it on a temporary branch with a m.MarshalXml() function to use temporarily.

**Comment by nadddy** [comment]
@clbanning 
Thanks, I will wait for it in master. I will keep scale tests disabled in my project till then. Thanks a lot for looking into this.

**Comment by clbanning** [comment]
Fixed at tip, v2.2.  m.Xml() now scales linearly as documented in README.
